pwndbg/commands/hijack_fd.py

```diff
@@ -0,0 +1,283 @@
+    from pwnlib import asm
+    from pwnlib import constants
+    from pwnlib import shellcraft
+    from pwnlib.util.net import sockaddr
+
+    sockdata, addr_len, _ = sockaddr(socket_data.address, socket_data.port, socket_data.ip_version)
+    socktype = {"tcp": "SOCK_STREAM", "udp": "SOCK_DGRAM"}[socket_data.protocol]
+    family = {"ipv4": "AF_INET", "ipv6": "AF_INET6"}[socket_data.ip_version]
+
+    regs = get_shellcode_regs()
+    stack_size = stack_size_alignment(len(sockdata))
+
+    dup_asm = (
+        shellcraft.syscall("SYS_dup2", regs["newfd"], replace_fd)
+        if hasattr(constants, "SYS_dup2")
+        else shellcraft.syscall("SYS_dup3", regs["newfd"], replace_fd, 0)
+    )
+
+    return stack_size, asm.asm(
+        "".join(
+            [
+                shellcraft.syscall("SYS_socket", family, socktype, 0),
+                shellcraft.mov(regs["newfd"], regs["syscall_ret"]),
+                shellcraft.pushstr(sockdata, False),
+                shellcraft.syscall("SYS_connect", regs["newfd"], regs["stack"], addr_len),
+                dup_asm,
+                shellcraft.syscall("SYS_close", regs["newfd"]),
+            ]
+        )
+    )
+
+
+@contextlib.asynccontextmanager
+async def exec_shellcode_with_stack(ec: pwndbg.dbg_mod.ExecutionController, blob, stack_size: int):
+    stack_start_diff = pwndbg.aglib.regs.sp
+    stack_start = stack_start_diff - stack_size
+    original_stack = pwndbg.aglib.memory.read(stack_start, stack_size)
+
+    try:
+        async with pwndbg.aglib.shellcode.exec_shellcode(
+            ec, blob, restore_context=True, disable_breakpoints=True
+        ):
+            stack_diff_size = stack_start_diff - pwndbg.aglib.regs.sp
+
+            # Make sure stack is not corrupted somehow
+            assert not (
+                stack_diff_size > stack_size
+            ), f"stack is probably corrupted size_current=f{stack_diff_size} size_max_want={stack_size}"
+
+            yield
+    finally:
+        pwndbg.aglib.memory.write(stack_start, original_stack)
+
+
+parser = argparse.ArgumentParser(
+    formatter_class=argparse.RawTextHelpFormatter,
+    description="""Replace file descriptors of a debugged process.
+
+The new file descriptor can point to:
+- a file
+- a pipe
+- a socket
+- a device, etc.
+
+Examples:
+1. Redirect STDOUT to a file:
+   `hijack-fd 1 /dev/null`
+
+2. Redirect STDERR to a socket:
+   `hijack-fd 2 tcp://localhost:8888`
+""",
+)
+
+parser.add_argument(
+    "fdnum",
+    help="File descriptor (FD) number to be replaced with the specified new socket or file.",
+    type=int,
+)
+
+
+class ParsedSocket(NamedTuple):
+    protocol: Literal["tcp", "udp"]
+    ip_version: Literal["ipv4", "ipv6"]
+    address: str
+    port: int
+
+
+def parse_socket(url: str) -> ParsedSocket:
+    if "://" in url:
+        parsed = urlparse(url)
+    else:
+        parsed = ParseResult("", url, "", "", "", "")
+
+    scheme_info = parsed.scheme.split("+", 1)
```

Comment:
What is the "+"? Why don't we have any example with it?

Reply:
I put examples here:
```py
parser.add_argument(
    "newfile",
    help="""Specify a file or a socket.
For files, the filename must start with `/` (e.g., `/etc/passwd`).
For sockets, the following formats are allowed:
- `127.0.0.1:80` (default is TCP)
- `tcp://[::1]:80`
- `udp://example.com:80`
- `tcp+ipv6://example.com:80`
    """,
    type=parse_file_or_socket,
)
```